File: src/js/window.js
// # The main window and its custom titlebar.

import { isMobile } from "./dom.js";
import { flushNoteSave, focusEditor, redoEditor, undoEditor } from "./editor.js";
import { shortcutActions } from "./shortcuts.js";
import { syncNow } from "./sync.js";
import {
  currentMonitor,
  getCurrentWindow,
  invoke,
  listen,
  PhysicalPosition,
  PhysicalSize,
} from "./tauri.js";

// ## Show the window, then put the caret in the editor.
// The window starts hidden to avoid a white flash, see the `show_window` command.
// The window state plugin has already put it back where it was by now, a first launch
// gets the centered window from tauri.conf.json instead.
export function setupMainWindow() {
  // Mobile has no window to place, and no `show_window` command.
  if (!isMobile) {
    invoke("show_window");

    // Pick up a restored fullscreen window, see `fullscreen` below.
    getCurrentWindow()
      .isFullscreen()
      .then((restored) => {
        fullscreen = restored;

        markFullscreen();
      });
  }

  focusEditor();
}

// ## Toggle the window between fullscreen and normal.
// Shared by the shortcut, the command palette, the menu item, and the macOS-style
// green dot. The window's own fullscreen report is unreliable on macOS, so the state
// lives here; every fullscreen path goes through this toggle, which keeps it in step.
// Startup is the exception, since nothing is in transition there and the plugin has to
// be asked what it restored.
let fullscreen = false;

export function toggleFullscreen() {
  // Mobile is always fullscreen.
  if (isMobile) return;

  fullscreen = !fullscreen;
  getCurrentWindow().setFullscreen(fullscreen);

  markFullscreen();
}

// ## Carry the fullscreen state over to the titlebar.
// The macOS-style green dot shows the way out of fullscreen while it is in it, exactly
// like the real one, which takes a hook for the CSS.
function markFullscreen() {
  document.getElementById("titlebar").classList.toggle("fullscreen", fullscreen);
}

// ## Let go of a fullscreen or maximized window, so a new placement can stick.
// macOS animates its way out of fullscreen and only lands on its old frame afterwards,
// so a placement set during that ride would be undone. Hence the wait, no event marks
// the landing, see `minimizeWindow` below for the same problem.
async function releaseWindow() {
  const currentWindow = getCurrentWindow();

  if (fullscreen) {
    toggleFullscreen();

    await new Promise((resolve) => setTimeout(resolve, 600));
  } else if (await currentWindow.isMaximized()) {
    await currentWindow.toggleMaximize();
  }
}

// ## Center the window on its screen.
export async function centerWindow() {
  // Mobile has no window to place.
  if (isMobile) return;

  await releaseWindow();

  getCurrentWindow().center();
}

// ## Move and resize the window to the left or right half of its screen.
// The monitor's work area is used rather than its full size, since that one leaves out
// the menu bar and the dock. Everything here is in physical pixels, which is what the
// monitor reports and what the window setters take.
// The size goes first: a macOS resize holds on to the window's bottom left corner and
// moves its top edge, so a position set before it would not survive the resize.
export async function tileWindow(toRight) {
  if (isMobile) return;

  const monitor = await currentMonitor();
  if (!monitor) return;

  await releaseWindow();

  const { position, size } = monitor.workArea;
  const width = Math.round(size.width / 2);

  const currentWindow = getCurrentWindow();

  await currentWindow.setSize(new PhysicalSize(width, size.height));

  await currentWindow.setPosition(
    new PhysicalPosition(toRight ? position.x + size.width - width : position.x, position.y),
  );
}

// ## Minimize the window, leaving fullscreen first when needed.
// macOS ignores a minimize while fullscreen and during the whole exit transition,
// and no event marks when the window has landed. So after leaving fullscreen the
// minimize is retried until the window reports it actually stuck.
async function minimizeWindow() {
  const currentWindow = getCurrentWindow();

  if (!fullscreen) {
    currentWindow.minimize();

    return;
  }

  toggleFullscreen();

  const deadline = Date.now() + 3000;
  while (Date.now() < deadline) {
    await new Promise((resolve) => setTimeout(resolve, 150));

    currentWindow.minimize();

    if (await currentWindow.isMinimized()) return;
  }
}

// ## Drag the titlebar out of fullscreen, and double click it to (un)maximize.
// The drag region cannot move a fullscreen window, so while fullscreen dragging it
// exits fullscreen instead. A plain double click always zooms the window bounds,
// like a native titlebar would, regardless of what the maximize button itself does.
function setupTitlebarDragAndDoubleClick(titlebar, currentWindow) {
  const isDragRegion = (event) =>
    event.button === 0 && event.target.hasAttribute("data-tauri-drag-region");

  titlebar.addEventListener("mousedown", (event) => {
    if (!fullscreen || !isDragRegion(event) || event.detail !== 1) return;

    // Only a real drag exits, a plain click (and its pixel of jitter) does nothing.
    const startX = event.screenX;
    const startY = event.screenY;

    const exitOnDrag = (move) => {
      if (Math.abs(move.screenX - startX) < 5 && Math.abs(move.screenY - startY) < 5) return;
      stopWatching();
      toggleFullscreen();
    };

    const stopWatching = () => {
      window.removeEventListener("mousemove", exitOnDrag);
      window.removeEventListener("mouseup", stopWatching);
    };

    window.addEventListener("mousemove", exitOnDrag);
    window.addEventListener("mouseup", stopWatching);
  });

  titlebar.addEventListener("dblclick", (event) => {
    if (!isDragRegion(event)) return;

    if (fullscreen) toggleFullscreen();
    else currentWindow.toggleMaximize();
  });
}

// ## The custom titlebar's window controls (the titlebar itself is hidden on mobile).
export function setupCustomTitlebar() {
  const currentWindow = getCurrentWindow();
  const titlebar = document.getElementById("titlebar");

  const actions = {
    minimize: minimizeWindow,
    // With the macOS-style controls the green dot goes fullscreen, like macOS does.
    // Either style leaves fullscreen on click, since toggleMaximize does nothing there.
    maximize: () =>
      fullscreen || titlebar.classList.contains("macos")
        ? toggleFullscreen()
        : currentWindow.toggleMaximize(),
    close: async () => {
      // Make sure a pending autosave isn't lost on the way out.
      await flushNoteSave();
      currentWindow.close();
    },
  };

  Object.entries(actions).forEach(([id, handler]) => {
    document.getElementById(`titlebar-${id}`)?.addEventListener("click", handler);
  });

  setupTitlebarDragAndDoubleClick(titlebar, currentWindow);
}

// ## Run an edit action where the focus is.
// The editor keeps its own undo history, so the webview's native undo cannot serve it.
// A focused text input still has nothing but the native one.
function runEditAction(editorAction, command) {
  if (document.activeElement?.closest("input, textarea")) {
    document.execCommand(command);
  } else {
    editorAction();
  }
}

// ## The macOS menu bar's custom items, arriving as events from the backend.
// Most ids match a shortcut action, so the menu and the shortcuts stay in step.
// The rest are the window items the undecorated window cannot serve natively.
export function setupMenuActions() {
  // The About panel's copyright year comes from here, since JS knows the local date
  // without any date math on the Rust side. About is for macOS only for now.
  // TODO - Maybe need to support an about for Linux later.
  if (!isMobile) invoke("menu_set_year", { year: new Date().getFullYear() });

  const menuOnlyActions = {
    minimize: minimizeWindow,
    zoom: () => getCurrentWindow().toggleMaximize(),
    "sync-now": () => syncNow(),
    undo: () => runEditAction(undoEditor, "undo"),
    redo: () => runEditAction(redoEditor, "redo"),
  };

  // ## Run the handler for the clicked menu item's id.
  listen("menuAction", ({ payload }) => {
    const action = menuOnlyActions[payload] ?? shortcutActions[payload]?.run;
    action?.();
  });
}

File: src-tauri/src/window.rs
// # Two patches over the window state plugin, both about where the window lands.
// One runs on the way in, the other on the way out, and both work on the plugin's own
// state file rather than on a copy of it.

use serde_json::Value;
use std::{fs, path::PathBuf};
use tauri::{AppHandle, Manager, PhysicalPosition, RunEvent};
use tauri_plugin_window_state::DEFAULT_FILENAME;

// ## The main window's label, which is also its key in the plugin's state file.
const MAIN_WINDOW: &str = "Minoo";

// ## The plugin's state file, and what it holds.
fn read_state(app: &AppHandle) -> Option<(PathBuf, Value)> {
    let path = app.path().app_config_dir().ok()?.join(DEFAULT_FILENAME);
    let state = serde_json::from_str(&fs::read_to_string(&path).ok()?).ok()?;

    Some((path, state))
}

// ## Put a restored window back on the corner it was saved on.
// The plugin restores the position before the size, and on macOS a resize keeps the
// window's bottom left corner, so the top edge slides down by whatever height the window
// just lost. A window smaller than the one in tauri.conf.json therefore comes back that
// much lower, and lower again on every launch, since the drifted corner is what gets
// saved next. Setting the position once more after the plugin is done fixes it: windows
// are built before this runs, and every frame change is queued on the main thread, so
// this one lands last.
pub fn restore_position(app: &AppHandle) {
    let Some((_, state)) = read_state(app) else {
        return;
    };

    let window_state = &state[MAIN_WINDOW];

    // A window filling the screen has no corner to correct, and already gets the size
    // from tauri.conf.json underneath it, see below.
    if window_state["fullscreen"] == true || window_state["maximized"] == true {
        return;
    }

    let (Some(x), Some(y)) = (window_state["x"].as_i64(), window_state["y"].as_i64()) else {
        return;
    };

    if let Some(window) = app.get_webview_window(MAIN_WINDOW) {
        let _ = window.set_position(PhysicalPosition {
            x: x as i32,
            y: y as i32,
        });
    }
}

// ## Hand a window that quit filling the screen its first launch placement back.
// The plugin records every resize it sees, and on macOS it cannot tell that an
// undecorated window is filling the screen, since the check it has for that is forced off
// there to work around `https://github.com/tauri-apps/tauri/issues/5812`. So quitting from
// fullscreen leaves the whole screen stored as the window's size, and the window has
// nowhere sane to land when it leaves fullscreen on the next launch.
//
// Such a window gets the placement of a first launch written over the plugin's own, so it
// lands centered and at the size from tauri.conf.json. Plugins get the exit event before
// the app itself does, so the last change can be made here.
pub fn reset_saved_placement(app: &AppHandle, event: &RunEvent) {
    if !matches!(event, RunEvent::Exit) {
        return;
    }

    let Some((path, mut state)) = read_state(app) else {
        return;
    };

    let Some(window_state) = state.get_mut(MAIN_WINDOW) else {
        return;
    };

    // A window that quit at a size of its own keeps whatever the plugin stored for it.
    if window_state["fullscreen"] != true && window_state["maximized"] != true {
        return;
    }

    // The monitor the window sat on, so a second screen keeps it.
    let x = window_state["x"].as_f64().unwrap_or_default();
    let y = window_state["y"].as_f64().unwrap_or_default();

    let Some(monitor) = app
        .monitor_from_point(x, y)
        .ok()
        .flatten()
        .or_else(|| app.primary_monitor().ok().flatten())
    else {
        return;
    };

    let Some(window_config) = app
        .config()
        .app
        .windows
        .iter()
        .find(|window| window.label == MAIN_WINDOW)
    else {
        return;
    };

    // The configured size is in logical pixels, the plugin stores physical ones.
    let width = (window_config.width * monitor.scale_factor()).round() as i32;
    let height = (window_config.height * monitor.scale_factor()).round() as i32;

    let left = monitor.position().x + (monitor.size().width as i32 - width) / 2;
    let top = monitor.position().y + (monitor.size().height as i32 - height) / 2;

    // A zoomed window is restored from `prev_x` and `prev_y`, a fullscreen one from `x`
    // and `y`, so both pairs get the same corner.
    for (key, value) in [
        ("x", left),
        ("prev_x", left),
        ("y", top),
        ("prev_y", top),
        ("width", width),
        ("height", height),
    ] {
        window_state[key] = value.into();
    }

    if let Ok(corrected) = serde_json::to_string(&state) {
        let _ = fs::write(&path, corrected);
    }
}
